faidx.c

```diff
@@ -360,55 +435,105 @@ int faidx_core(int argc, char *argv[], enum fai_format_options format)
+            hts_opt *opts = (hts_opt *)(out.gopt->out.specific);
+            char mode[13] = "w";
+            int level = 0;
+            while (opts) {
+                if (opts->opt == HTS_OPT_COMPRESSION_LEVEL) {   //compression level
+                    level = opts->val.i;
+                    break;
+                }
+                opts = opts->next;
+            }
+            if (level > 0) {
+                snprintf(mode, sizeof(mode), "w%d", level);     //pass compression with mode
+            }
+            out.bgzf_fp = bgzf_open(output_file, mode);
         }
 
-        file_out = fopen( output_file, "w" );
-
-        if( file_out == NULL) {
+        if( (!out.isbgzip && out.fp == NULL) || (out.isbgzip && out.bgzf_fp == NULL)) {
             fprintf(stderr,"[faidx] Cannot open \"%s\" for writing :%s.\n", output_file, strerror(errno) );
-            return EXIT_FAILURE;
+            goto exit2;
+        }
+        //set threading if set
+        if (out.isbgzip && out.gopt->nthreads > 0) {
+            if (bgzf_mt(out.bgzf_fp, out.gopt->nthreads, 256/*unused*/) < 0) {
+                fprintf(stderr,"[faidx] Failed to set threads\n");
+                goto exit1;
+            }
         }
     }
 
-    int exit_status = EXIT_SUCCESS;
-
     if (region_file) {
         hFILE *rf;
 
         if ((rf = hopen(region_file, "r"))) {
-            exit_status = read_regions_from_file(fai, rf, file_out, ignore_error, line_len, rev, pos_strand_name, neg_strand_name, format);
+            exit_status = read_regions_from_file(fai, rf, &out, ignore_error, line_len, rev, pos_strand_name, neg_strand_name, format);
 
             if (hclose(rf) != 0) {
                 fprintf(stderr, "[faidx] Warning: failed to close %s", region_file);
             }
+            if (exit_status == EXIT_FAILURE) {
+                goto exit1;
+            }
         } else {
             fprintf(stderr, "[faidx] Failed to open \"%s\" for reading.\n", region_file);
-            exit_status = EXIT_FAILURE;
+            goto exit1;
         }
     }
 
+    exit_status = EXIT_SUCCESS;
     while ( ++optind<argc && exit_status == EXIT_SUCCESS) {
-        exit_status = write_output(fai, file_out, argv[optind], ignore_error, line_len, rev, pos_strand_name, neg_strand_name, format);
+        exit_status = write_output(fai, &out, argv[optind], ignore_error, line_len, rev, pos_strand_name, neg_strand_name, format);
     }
 
-    fai_destroy(fai);
-
-    if (fflush(file_out) == EOF) {
-        print_error_errno("faidx", "failed to flush output");
-        exit_status = EXIT_FAILURE;
+    if (!out.isbgzip) {
+        if (fflush(out.fp) == EOF) {
+            print_error_errno("faidx", "Failed to flush output\n");
+            exit_status = EXIT_FAILURE;
+        }
+    }
+    else {
+        if (bgzf_flush(out.bgzf_fp) == EOF) {
+            print_error_errno("faidx", "Failed to flush output\n");
+            exit_status = EXIT_FAILURE;
+        }
+    }
```

Comment:
Could be more succint and remove code duplication.  Eg

```
int flushed = out.isbgzip
    ? fflush(out.fp)
    : bgzf_flush(out.bgzf_fp);
if (flushed == EOF) {
    print_error_errno("faidx", "Failed to flush output\n");
    exit_status = EXIT_FAILURE;
}
```

Reply:
Done.